pandas/_libs/groupby.pyx

```diff
@@ -412,16 +414,20 @@ def group_any_all(uint8_t[::1] out,
         String object dictating whether to use any or all truth testing
     skipna : bool
         Flag to ignore nan values during truth testing
+    masked : bool
+        If True, compute the result using Kleene logic
 
     Notes
     -----
     This method modifies the `out` parameter rather than returning an object.
-    The returned values will either be 0 or 1 (False or True, respectively).
+    The returned values will either be 0, 1 (False or True, respectively), or
+    2 to signify a masked position in the case of a nullable input.
```

Comment:
hmm is this what we do elsewhere to encode Kleene logic? (e.g. would rather use -1 on nulls)

Reply:
The other algo is a bit different since it can be vectorized and computes both data and mask (so doesn't need to store something signifying NULL). Agree that -1 is more intuitive, have changed to use -1 as the mask signal

Reply:
At some point, it might be interesting to look into directly filling in the mask as well instead of using the temporary -1.

Personally, I would maybe leave it at using 2, so you don't need to change the uint8 type

Reply:
Yep I don't think that would be too bad, that's how I had approached it originally, but changes were slightly more invasive, so stuck with this.

For the -1 vs 2 question, I think I slightly prefer -1 because it seems more immediately apparent what it represents. Although it forces changing the dtype, still the same number of bytes, so memory layout is unaffected and constructing the view looks equivalent speed-wise:

```
In [1]: import numpy as np

In [2]: arr = np.array([True, False, True])

In [3]: %timeit arr.view('uint8')
332 ns ± 2.84 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit arr.view('int8')
334 ns ± 3.78 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
But not a strong opinion, fine with using 2 as well :)

Reply:
If speed-wise the int8 vs uint8 doesn't matter, then indeed the -1 is fine as well